bundles/org.openhab.binding.mycroft/src/main/java/org/openhab/binding/mycroft/internal/MycroftHandler.java

```diff
@@ -0,0 +1,234 @@
+import org.openhab.binding.mycroft.internal.channels.VolumeChannel;
+import org.openhab.core.io.net.http.WebSocketFactory;
+import org.openhab.core.thing.Channel;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.State;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link MycroftHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Gwendal ROULLEAU - Initial contribution
+ */
+@NonNullByDefault
+public class MycroftHandler extends BaseThingHandler implements MycroftConnectionListener {
+
+    private final Logger logger = LoggerFactory.getLogger(MycroftHandler.class);
+
+    protected final MycroftConnection connection;
+    private @Nullable ScheduledFuture<?> scheduledFuture;
+    private MycroftConfiguration config = new MycroftConfiguration();
+    private boolean thingDisposing = false;
+    protected Map<ChannelUID, MycroftChannel<?>> mycroftChannels = new HashMap<>();
+
+    /** The reconnect frequency in case of error */
+    private static final int POLL_FREQUENCY_SEC = 30;
+    private int sometimesSendVolumeRequest = 0;
+
+    public MycroftHandler(Thing thing, WebSocketFactory webSocketFactory) {
+        super(thing);
+        String websocketID = thing.getUID().getAsString().replace(':', '-');
+        if (websocketID.length() < 4) {
+            websocketID = "mycroft-" + websocketID;
+        }
+        if (websocketID.length() > 20) {
+            websocketID = websocketID.substring(websocketID.length() - 20);
+        }
+        this.connection = new MycroftConnection(this, webSocketFactory.createWebSocketClient(websocketID));
+    }
+
+    /**
+     * Stops the API request or websocket reconnect timer
+     */
+    private void stopTimer() {
+        ScheduledFuture<?> future = scheduledFuture;
+        if (future != null) {
+            future.cancel(false);
+            scheduledFuture = null;
+        }
+    }
+
+    /**
+     * Starts the websocket connection.
+     * sometimes send a get volume request to fully test the connection / refresh volume.
+     */
+    private void startWebsocket() {
+        if (thingDisposing) {
+            return;
+        }
+        if (connection.isConnected()) {
+            // sometimes test the connection by sending a real message
+            // AND refreshing volume in the same step
+            if (sometimesSendVolumeRequest >= 3) { // arbitrary one on three times
+                sometimesSendVolumeRequest = 0;
+                sendMessage(new MessageVolumeGet());
+            } else {
+                sometimesSendVolumeRequest++;
+            }
+        } else {
+            connection.start(config.host, config.port);
```

Comment:
What will happen if the host cannot be reached? Will it result in a `connectionLost()`call, even if it had never been established?

Reply:
Yes, if the host cannot be reached, it will result in a connectionLost() call, even with no previous connection established.
And then the thing will be updated.
During my test this behaviour was very consistant : the thing status is solid, can be trusted, and provide usefull information.